cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,608 @@
+# Plugable Cloud Provider over gRPC
+
+
+Authors:
+
+* Hector Fernandez <<hfernandez@d2iq.com>> (@hectorj2f)
+
+
+## Motivation
+
+CA is released as a bundle which includes a hardcoded list of supported cloud providers.
+Whenever users want to implement the logic of their own cloud provider, they need to fork the CA and add their own implementation.
+
+In particular users need to follow these steps to support a custom private cloud:
+
+* Write a client for your private cloud in Go, implementing CloudProvider interface.
+
+* Add constructing it to cloud provider builder.
+
+* Build a custom image of Cluster Autoscaler that includes those changes and configure it to start with your cloud provider.
+
+This is a concern that has been raised in the past [PR953](https://github.com/kubernetes/autoscaler/issues/953) and [PR1060](https://github.com/kubernetes/autoscaler/issues/1060).
+
+Therefore a new implemetation should be added to CA in order to extend it without breaking any backwards compatibility or
+the current cloud provider implementations.
+
+## Goals
+
+* Support custom cloud provider implementations without changing the current `CloudProvider` interface.
+* Make CA extendable, so users do not need to fork the CA repository.
+
+## Proposal
+
+There are couple of examples of pluggable designs using Go SDKs that would guide us on how to extend CA
+to support custom providers as plugins.
+This approach is inspired based on [Hashicorp go-plugin](https://github.com/hashicorp/go-plugin) and [Grafana Go SDK for plugins](https://github.com/grafana/grafana-plugin-sdk-go).
+
```

Comment:
i think it would be nice to add something here about the general architecture that is being proposed, just to help the reader understand better. as i understand it, regardless of the option chosen, this proposal will involve the CA and  a provider-specific gRPC server running as a separate binary.

Reply:
yes, that is correct. I'll add a graph to the kep.